# Анализ графика движения транспортных средств на маршруте города Москвы
#### Треубется проанализировать график движения транспортных средств на маршруте: рассчитать среднее время движения ТС на маршруте, определить участок маршрута, на котором наибольшая задержка, а также выработать метрику, показывающее отклонение фактического времени движения от графика.

## Импорт данных и предварительная обработка

In [11]:
import pandas as pd

In [12]:
shedule_plan = pd.read_excel(r"C:\Users\Turner\Documents\Тетрадки\schedule_plan.xlsx")
shedule_fact = pd.read_excel(r"C:\Users\Turner\Documents\Тетрадки\schedule_fact.xlsx")

C:\Program Files (x86)\Anaconda\lib\site-packages\openpyxl\styles\stylesheet.py:226: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")
C:\Program Files (x86)\Anaconda\lib\site-packages\openpyxl\styles\stylesheet.py:226: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")


In [13]:
display(shedule_plan.head(20))
display(shedule_fact.head(20))

date  route_id  grafic  trip_num  stop_sequence  trip_id  stop_id  \
0  2023-02-06       988     301         1              1    67765  1009938   
1  2023-02-06       988     301         1              2    67765  1009937   
2  2023-02-06       988     301         1              3    67765    13835   
3  2023-02-06       988     301         1              4    67765    13834   
4  2023-02-06       988     301         1              5    67765    13832   
5  2023-02-06       988     301         1              6    67765    13952   
6  2023-02-06       988     301         1              7    67765    13951   
7  2023-02-06       988     301         1              8    67765    13949   
8  2023-02-06       988     301         1              9    67765    13947   
9  2023-02-06       988     301         1             10    67765    13942   
10 2023-02-06       988     301         1             11    67765    13943   
11 2023-02-06       988     301         1             12    67765    13945   
12 2023-02-06       988     301         1             13    67765    13941   
13 2023-02-06       988     301         1             14    67765    13788   
14 2023-02-06       988     301         1             15    67765  1009938   
15 2023-02-06       988     301         2              1    67765  1009938   
16 2023-02-06       988     301         2              2    67765  1009937   
17 2023-02-06       988     301         2              3    67765    13835   
18 2023-02-06       988     301         2              4    67765    13834   
19 2023-02-06       988     301         2              5    67765    13832   

          arrival_time      departure_time  
0  2023-02-06 05:35:00 2023-02-06 05:35:00  
1  2023-02-06 05:35:00 2023-02-06 05:35:00  
2  2023-02-06 05:37:00 2023-02-06 05:37:00  
3  2023-02-06 05:39:00 2023-02-06 05:39:00  
4  2023-02-06 05:42:00 2023-02-06 05:42:00  
5  2023-02-06 05:45:00 2023-02-06 05:45:00  
6  2023-02-06 05:47:00 2023-02-06 05:47:00  
7  2023-02-06 05:50:00 2023-02-06 05:50:00  
8  2023-02-06 05:51:00 2023-02-06 05:51:00  
9  2023-02-06 05:52:00 2023-02-06 05:52:00  
10 2023-02-06 05:53:00 2023-02-06 05:53:00  
11 2023-02-06 05:54:00 2023-02-06 05:54:00  
12 2023-02-06 05:55:00 2023-02-06 05:55:00  
13 2023-02-06 05:56:00 2023-02-06 05:56:00  
14 2023-02-06 05:56:00 2023-02-06 05:56:00  
15 2023-02-06 06:00:00 2023-02-06 06:00:00  
16 2023-02-06 06:00:00 2023-02-06 06:00:00  
17 2023-02-06 06:02:00 2023-02-06 06:02:00  
18 2023-02-06 06:04:00 2023-02-06 06:04:00  
19 2023-02-06 06:07:00 2023-02-06 06:07:00

date  route_id  grafic  trip_num  stop_sequence  trip_id  stop_id  \
0  2023-02-06       988     301        -2              1    67765  1009938   
1  2023-02-06       988     301        -2              2    67765  1009937   
2  2023-02-06       988     301        -2             10    67765    13942   
3  2023-02-06       988     301        -2             11    67765    13943   
4  2023-02-06       988     301        -2             12    67765    13945   
5  2023-02-06       988     301        -2             13    67765    13941   
6  2023-02-06       988     301        -2             14    67765    13788   
7  2023-02-06       988     301        -2             15    67765  1009938   
8  2023-02-06       988     301        -1              1    67765  1009938   
9  2023-02-06       988     301        -1              2    67765  1009937   
10 2023-02-06       988     301        -1             11    67765    13943   
11 2023-02-06       988     301        -1             12    67765    13945   
12 2023-02-06       988     301        -1             13    67765    13941   
13 2023-02-06       988     301        -1             14    67765    13788   
14 2023-02-06       988     301        -1             15    67765  1009938   
15 2023-02-06       988     301         0              1    67765  1009938   
16 2023-02-06       988     301         0              2    67765  1009937   
17 2023-02-06       988     301         1              3    67765    13835   
18 2023-02-06       988     301         1              4    67765    13834   
19 2023-02-06       988     301         1              5    67765    13832   

          arrival_time      departure_time  
0  2023-02-06 19:01:14 2023-02-06 19:01:21  
1  2023-02-06 19:01:29 2023-02-06 19:01:33  
2  2023-02-06 19:53:07 2023-02-06 19:53:11  
3  2023-02-06 19:53:41 2023-02-06 19:53:45  
4  2023-02-06 19:54:29 2023-02-06 19:54:33  
5  2023-02-06 19:55:10 2023-02-06 19:55:13  
6  2023-02-06 19:55:59 2023-02-06 19:56:04  
7  2023-02-06 19:56:24 2023-02-06 19:56:33  
8  2023-02-06 09:04:00 2023-02-06 09:04:04  
9  2023-02-06 09:04:14 2023-02-06 09:04:18  
10 2023-02-06 10:03:59 2023-02-06 10:04:03  
11 2023-02-06 10:04:32 2023-02-06 10:04:35  
12 2023-02-06 10:05:09 2023-02-06 10:05:14  
13 2023-02-06 10:05:58 2023-02-06 10:06:03  
14 2023-02-06 10:06:13 2023-02-06 10:06:18  
15 2023-02-06 22:57:49 2023-02-06 22:57:53  
16 2023-02-06 22:58:02 2023-02-06 22:58:06  
17 2023-02-06 05:38:00 2023-02-06 05:38:20  
18 2023-02-06 05:39:42 2023-02-06 05:39:58  
19 2023-02-06 05:42:02 2023-02-06 05:42:08

#### Приведение дат к типу datetime

In [14]:
shedule_plan['date'] = pd.to_datetime(shedule_plan['date'])
shedule_plan['arrival_time'] = pd.to_datetime(shedule_plan['arrival_time'])
shedule_plan['departure_time'] = pd.to_datetime(shedule_plan['departure_time'])

shedule_fact['date'] = pd.to_datetime(shedule_fact['date'])
shedule_fact['arrival_time'] = pd.to_datetime(shedule_fact['arrival_time'])
shedule_fact['departure_time'] = pd.to_datetime(shedule_fact['departure_time'])

#### Поиск пропусков и дубликатов

In [15]:
shedule_plan.isna().value_counts()

date   route_id  grafic  trip_num  stop_sequence  trip_id  stop_id  arrival_time  departure_time
False  False     False   False     False          False    False    False         False             450
dtype: int64

In [16]:
shedule_plan.duplicated().value_counts()

False    450
dtype: int64

#### Количество маршрутов

In [17]:
print(shedule_fact['route_id'].nunique())
print(shedule_plan['route_id'].nunique())

1
1


#### Количество рейсов

In [18]:
trip_num = shedule_fact['trip_num'].unique()

print(shedule_fact['trip_num'].nunique())
print(shedule_plan['trip_num'].nunique())

14
10


#### Количество остановок

In [19]:
print(shedule_fact['stop_id'].nunique())
print(shedule_plan['stop_id'].nunique())

14
14


In [20]:
print(shedule_fact['stop_sequence'].nunique())
print(shedule_plan['stop_sequence'].nunique())

15
15


#### Количество ТС

In [21]:
machine = shedule_fact['grafic'].unique()
print(machine)
print(shedule_fact['grafic'].nunique())
print(shedule_plan['grafic'].nunique())

[301 702 703]
3
3


### Расчёт среднего времени

In [22]:
times_trip = '0 days 00:00:00'
n = 0

for m in machine:
    for r in trip_num:
        shed_fact_cur = shedule_fact.query('trip_num == @r & grafic == @m')
        final_stop = shed_fact_cur['stop_sequence'].max()
        start_stop = shed_fact_cur['stop_sequence'].min()
        if final_stop == 15 and start_stop == 1:
            t2 = shed_fact_cur.loc[shed_fact_cur['stop_sequence'] == final_stop, 'arrival_time'].tolist()
            t1 = shed_fact_cur.loc[shed_fact_cur['stop_sequence'] == start_stop, 'departure_time'].tolist()
            delta = t2[0] - t1[0]
            times_trip = times_trip + delta
            n = n+1

print(times_trip)
print(n)

0 days 19:52:21
33


In [23]:
print('Среднее время следования', (times_trip/n))

Среднее время следования 0 days 00:36:07.909090909


### Определение участка маршрута, на котором наибольшая задержка.

#### Объединение таблиц shedule_plan и shedule_fact.

In [24]:
fact_plan = shedule_plan.merge(shedule_fact, on=['date', 
                                                 'route_id', 
                                                 'grafic', 
                                                 'trip_num', 
                                                 'stop_id', 
                                                 'stop_sequence', 
                                                 'trip_id'], 
                               how='inner')

fact_plan = fact_plan.rename(columns={'arrival_time_x' : 'arrival_plan',
                                      'arrival_time_y' : 'arrival_fact',
                                      'departure_time_x' : 'departure_plan',
                                      'departure_time_y' : 'departure_fact'})

display(fact_plan.head())

date  route_id  grafic  trip_num  stop_sequence  trip_id  stop_id  \
0 2023-02-06       988     301         1              3    67765    13835   
1 2023-02-06       988     301         1              4    67765    13834   
2 2023-02-06       988     301         1              5    67765    13832   
3 2023-02-06       988     301         1              6    67765    13952   
4 2023-02-06       988     301         1              7    67765    13951   

         arrival_plan      departure_plan        arrival_fact  \
0 2023-02-06 05:37:00 2023-02-06 05:37:00 2023-02-06 05:38:00   
1 2023-02-06 05:39:00 2023-02-06 05:39:00 2023-02-06 05:39:42   
2 2023-02-06 05:42:00 2023-02-06 05:42:00 2023-02-06 05:42:02   
3 2023-02-06 05:45:00 2023-02-06 05:45:00 2023-02-06 05:44:21   
4 2023-02-06 05:47:00 2023-02-06 05:47:00 2023-02-06 05:47:00   

       departure_fact  
0 2023-02-06 05:38:20  
1 2023-02-06 05:39:58  
2 2023-02-06 05:42:08  
3 2023-02-06 05:44:31  
4 2023-02-06 05:47:09

#### Добавим столбец с временем задержки по отправлению

In [25]:
fact_plan['delay'] = fact_plan['departure_fact'] - fact_plan['departure_plan']
display(fact_plan.head())

date  route_id  grafic  trip_num  stop_sequence  trip_id  stop_id  \
0 2023-02-06       988     301         1              3    67765    13835   
1 2023-02-06       988     301         1              4    67765    13834   
2 2023-02-06       988     301         1              5    67765    13832   
3 2023-02-06       988     301         1              6    67765    13952   
4 2023-02-06       988     301         1              7    67765    13951   

         arrival_plan      departure_plan        arrival_fact  \
0 2023-02-06 05:37:00 2023-02-06 05:37:00 2023-02-06 05:38:00   
1 2023-02-06 05:39:00 2023-02-06 05:39:00 2023-02-06 05:39:42   
2 2023-02-06 05:42:00 2023-02-06 05:42:00 2023-02-06 05:42:02   
3 2023-02-06 05:45:00 2023-02-06 05:45:00 2023-02-06 05:44:21   
4 2023-02-06 05:47:00 2023-02-06 05:47:00 2023-02-06 05:47:00   

       departure_fact             delay  
0 2023-02-06 05:38:20   0 days 00:01:20  
1 2023-02-06 05:39:58   0 days 00:00:58  
2 2023-02-06 05:42:08   0 days 00:00:08  
3 2023-02-06 05:44:31 -1 days +23:59:31  
4 2023-02-06 05:47:09   0 days 00:00:09

In [26]:
stages = fact_plan.groupby('stop_sequence', as_index=False)['delay'].mean()
display(stages)

stop_sequence                       delay
0               1 -1 days +23:56:42.321428572
1               2 -1 days +23:57:59.724137932
2               3   0 days 00:01:52.166666666
3               4   0 days 00:01:16.166666666
4               5 -1 days +23:59:56.133333334
5               6 -1 days +23:59:53.333333334
6               7   0 days 00:00:21.866666666
7               8 -1 days +23:59:15.133333334
8               9    -1 days +23:59:28.900000
9              10 -1 days +23:59:31.366666667
10             11 -1 days +23:59:38.566666667
11             12    -1 days +23:59:43.100000
12             13    -1 days +23:59:59.100000
13             14   0 days 00:00:37.233333333
14             15   0 days 00:00:56.333333333

In [27]:
display(stages.loc[stages['delay'] == stages['delay'].max(),:])

stop_sequence                     delay
2              3 0 days 00:01:52.166666666

#### Наибольшее опоздание накапливается на перегоне между остановками с порядковыми номерами 2 и 3. Выведем их идентификаторы.

In [36]:
print(fact_plan.loc[fact_plan['stop_sequence'] == 2,'stop_id'].unique())
print(fact_plan.loc[fact_plan['stop_sequence'] == 3,'stop_id'].unique())

[1009937]
[13835]


#### Соответсвующие номера остановок - 1009937 и 13835.

### Метрика пунктуальности для каждого из трёх ТС.
#### Среднее куммулятивное отклонение за один рейс - накопленное абсолютное отклонение от расписания по всем остановкам за рейс и усреднённое по всем рейсам каждого ТС.

In [19]:
# расчёт абсолютного отклонения

def absolute(value):
    t0 = pd.Timedelta(value='0 days 00:00:00')
    if value < t0:
        return (t0 - value)
    else:
        return value

fact_plan['abs_dev'] = fact_plan['delay'].apply(absolute)
display(fact_plan.head())

date  route_id  grafic  trip_num  stop_sequence  trip_id  stop_id  \
0 2023-02-06       988     301         1              3    67765    13835   
1 2023-02-06       988     301         1              4    67765    13834   
2 2023-02-06       988     301         1              5    67765    13832   
3 2023-02-06       988     301         1              6    67765    13952   
4 2023-02-06       988     301         1              7    67765    13951   

         arrival_plan      departure_plan        arrival_fact  \
0 2023-02-06 05:37:00 2023-02-06 05:37:00 2023-02-06 05:38:00   
1 2023-02-06 05:39:00 2023-02-06 05:39:00 2023-02-06 05:39:42   
2 2023-02-06 05:42:00 2023-02-06 05:42:00 2023-02-06 05:42:02   
3 2023-02-06 05:45:00 2023-02-06 05:45:00 2023-02-06 05:44:21   
4 2023-02-06 05:47:00 2023-02-06 05:47:00 2023-02-06 05:47:00   

       departure_fact             delay         abs_dev  
0 2023-02-06 05:38:20   0 days 00:01:20 0 days 00:01:20  
1 2023-02-06 05:39:58   0 days 00:00:58 0 days 00:00:58  
2 2023-02-06 05:42:08   0 days 00:00:08 0 days 00:00:08  
3 2023-02-06 05:44:31 -1 days +23:59:31 0 days 00:00:29  
4 2023-02-06 05:47:09   0 days 00:00:09 0 days 00:00:09

#### Суммарное абсолютное отклонение для всех ТС и рейсов.

In [20]:
abs_group = fact_plan.groupby(by=['grafic', 'trip_num'], as_index=False)['abs_dev'].sum()
display(abs_group.head())

grafic  trip_num         abs_dev
0     301         1 0 days 00:08:02
1     301         2 0 days 00:13:57
2     301         3 0 days 00:11:51
3     301         4 0 days 00:16:39
4     301         5 0 days 00:23:00

#### Абсолютное куммулятивное отклонение за рейс, усреднённое по количеству выходов для каждого маршрута.

In [21]:
abs_group_2 = abs_group.groupby(by='grafic', as_index=False)['abs_dev'].mean()
display(abs_group_2)

grafic                abs_dev
0     301 0 days 00:19:53.600000
1     702 0 days 00:24:18.800000
2     703 0 days 00:23:23.300000